src/openrct2/actions/SmallSceneryPlaceAction.hpp

```diff
@@ -0,0 +1,422 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2019 OpenRCT2 developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#pragma once
+
+#include "../Cheats.h"
+#include "../OpenRCT2.h"
+#include "../common.h"
+#include "../core/MemoryStream.h"
+#include "../interface/Window.h"
+#include "../localisation/Localisation.h"
+#include "../localisation/StringIds.h"
+#include "../management/Finance.h"
+#include "../ride/Ride.h"
+#include "../ride/TrackDesign.h"
+#include "../world/MapAnimation.h"
+#include "../world/Park.h"
+#include "../world/SmallScenery.h"
+#include "../world/Sprite.h"
+#include "../world/Surface.h"
+#include "../world/TileElement.h"
+#include "GameAction.h"
+
+DEFINE_GAME_ACTION(SmallSceneryPlaceAction, GAME_COMMAND_PLACE_SCENERY, GameActionResult)
+{
+private:
+    CoordsXYZD _loc;
+    uint8_t _quadrant;
+    uint8_t _sceneryType;
+    uint8_t _primaryColour;
+    uint8_t _secondaryColour;
+
+public:
+    SmallSceneryPlaceAction() = default;
+
+    SmallSceneryPlaceAction(
+        CoordsXYZD loc, uint8_t quadrant, uint8_t sceneryType, uint8_t primaryColour, uint8_t secondaryColour)
+        : _loc(loc)
+        , _quadrant(quadrant)
+        , _sceneryType(sceneryType)
+        , _primaryColour(primaryColour)
+        , _secondaryColour(secondaryColour)
+    {
+    }
+
+    uint16_t GetActionFlags() const override
+    {
+        return GameAction::GetActionFlags();
+    }
+
+    void Serialise(DataSerialiser & stream) override
+    {
+        GameAction::Serialise(stream);
+
+        stream << DS_TAG(_loc) << DS_TAG(_quadrant) << DS_TAG(_sceneryType) << DS_TAG(_primaryColour)
+               << DS_TAG(_secondaryColour);
+    }
+
+    GameActionResult::Ptr Query() const override
+    {
+        bool isOnWater = false;
+        bool supportsRequired = false;
+        if (_loc.z != 0)
+        {
+            supportsRequired = true;
+        }
+        int32_t baseHeight = tile_element_height(_loc.x, _loc.y);
+        // If on water
+        if (baseHeight & 0xFFFF0000)
+        {
+            baseHeight >>= 16;
+        }
+        auto res = MakeResult();
+        res->Position.x = _loc.x + 16;
+        res->Position.y = _loc.y + 16;
+        res->Position.z = baseHeight;
+        if (_loc.z != 0)
+        {
+            baseHeight = _loc.z;
+            res->Position.z = baseHeight;
+        }
+
+        if (!map_check_free_elements_and_reorganise(1))
+        {
+            return MakeResult(GA_ERROR::NO_FREE_ELEMENTS, STR_CANT_POSITION_THIS_HERE);
+        }
+
+        if (!byte_9D8150 && (_loc.x > gMapSizeMaxXY || _loc.y > gMapSizeMaxXY))
```

Comment:
I believe byte_9D8150 is set when it wants the preview only. We could call this something like 'previewMode' but thats better for a later PR.

Reply:
Yes that is correct. I thought about that as well. Really it should be passed as a game command flag but I'll wait until we are doing the track place action to change it.